tests/mainnet/test_getters.py

```diff
@@ -0,0 +1,726 @@
+import itertools
+import warnings
+
+import brownie
+import pytest
+
+from .abis import curve_pool, curve_pool_v2, gauge_controller
+from .utils.constants import (
+    METAREGISTRY_CRYPTO_FACTORY_HANDLER_INDEX,
+    METAREGISTRY_CRYPTO_REGISTRY_HANDLER_INDEX,
+    METAREGISTRY_STABLE_FACTORY_HANDLER_INDEX,
+    METAREGISTRY_STABLE_REGISTRY_HANDLER_INDEX,
+)
+
+MAX_REGISTRIES = 4
+MAX_POOLS_TO_CHECK = 200
+
+
+# --------- Helper Functions ---------
+
+
+def get_coin_decimals_from_pool(pool, curve_api):
+    """Gets coin decimals for a pool.
+
+    Note: this will not get underlying decimals for metapools.
+
+    Args:
+        pool (str): pool address
+        curve_api (fixture): Fixture of a simple CurveAPI mock contract
+
+    Returns:
+        tuple (length 4): coin decimals
+    """
+    coin_decimals = []
+    for coin_id in range(4):
+        try:
+            coin_decimals.append(curve_api.get_decimals_for_coin_in_pool(pool, coin_id))
+        except brownie.exceptions.VirtualMachineError:
+            break
+    return coin_decimals
+
+
+def get_registry_and_pool(registries, registry_id, max_pools, pool_index, registry_pool_count):
+    """Returns registry and pool for a given registry and pool index.
+
+    Args:
+        registries (list): list of child registry brownie.Contract instances
+        registry_id (int): metaregistry index of the registry
+        max_pools (int): max pools to test per registry. Set in conftest.
+        pool_index (int): index of pool in registry.
+
+    Returns:
+        tuple(brownie.Contract, address): registry and pool.
+    """
+
+    registry = registries[registry_id]
+    total_pools = registry_pool_count[registry] if max_pools == 0 else max_pools
+    if pool_index > (total_pools - 1):
+        pytest.skip(f"`pool_index` {pool_index} > `total_pool` {total_pools}")
+    pool = registry.pool_list(pool_index)
+    return registry, pool
+
+
+def check_pool_already_registered(metaregistry, pool, registry):
+    """Checks whether the pool was already registered. Skips test if it was.
+
+    Args:
+        metaregistry (fixture): MetaRegistry contract fixture.
+        pool (str): address of pool
+        registry (brownie.Contract): Contract instance of registry.
+    """
+    registry_in_meta_registry = metaregistry.get_registry_handler_from_pool(pool)
+    if registry != registry_in_meta_registry:
+        pytest.skip(
+            "Pool already registred in another registry. "
+            f"Pool: {pool}, "
+            f"registry: {registry}, "
+            f"metaregistered registry: {registry_in_meta_registry}"
+        )
+
+
+# --------- Parameterised Tests: checkers ---------
+
+
+@pytest.mark.parametrize("registry_id", list(range(MAX_REGISTRIES)))
+@pytest.mark.parametrize("pool_index", list(range(MAX_POOLS_TO_CHECK)))
```

Comment:
Why would you use `MAX_POOLS_TO_CHECK` here instead of the actual number of pools?
If `MAX_POOLS_TO_CHECK` is higher than the actual number of pools, you're still wasting time running the test, checking that there's no pool in `get_registry_and_pool()` and then skipping the test.
If `MAX_POOLS_TO_CHECK` is lower than the actual number of pools, you're not testing on all the pools.

Reply:
I think it's fine to parametrize for the registries, but I'd avoid using it to loop through the pool count. You'll be initializing tests for every single pool and that will take forever. 4 registries * 200 pools and you're up for 800 tests. And then half of those get skipped anyways (but still take time to set up!). Much better to use max_pools as a fixture and a loop in your test.

As a comparison, parametrizing over the range of MAX_POOLS, it took over 6 minutes to run the tests for the very basic `test_is_registered`:
![image](https://user-images.githubusercontent.com/25791237/174688651-66470979-331d-4ffa-ae7f-0e5de80369dd.png)

Disabling the parametrization and using `for pool_index in range(max_pools):`, the test runs in 30 seconds
![image](https://user-images.githubusercontent.com/25791237/174688571-c4941cfe-5ee5-46d5-83b3-98ebb00b42d8.png)

Reply:
I guess you could still:
- Keep the pool list for each registry as a fixture so you only have to make all these calls once at runtime (as opposed to once every test)
- Use multicall in the fixture to cut down the number of calls you're making to each registry.
- Use multicall in the test to batch the calls to `metaregistry.is_registered`

You'll need to loop inside the test for multicall (so no parametrization on `MAX_POOL_COUNT`)

Reply:
using loop without multicall (since we discussed that even there the results are similar)